pkg/controller/node/nodecontroller.go

```diff
@@ -408,14 +408,13 @@ func (nc *NodeController) Run() {
 // post "NodeReady==ConditionUnknown". It also evicts all pods if node is not ready or
 // not reachable for a long period of time.
 func (nc *NodeController) monitorNodeStatus() error {
-	// It is enough to list Nodes from apiserver, since we can tolerate some small
-	// delays comparing to state from etcd and there is eventual consistency anyway.
-	// TODO: We should list them from local cache: nodeStore.
-	nodes, err := nc.kubeClient.Core().Nodes().List(v1.ListOptions{ResourceVersion: "0"})
+	// We are listing nodes from local cache as we can tolerate some small delays
+	// comparing to state from etcd and there is eventual consistency anyway.
+	nodes, err := nc.nodeStore.List()
```

Comment:
There is a GET in L453, you probably want to switch that to a cache lookup too. Also it seems that you want to deep-copy in certain places inside `tryUpdateNodeStatus`.